pkg/config/expand.go

```diff
@@ -186,6 +187,33 @@ func (bp Blueprint) expandBackend(grp *Group) {
 	}
 }
 
+func kubectlProviderRequiredModule(grp *Group) (bool, Module) {
+	for _, mod := range grp.Modules {
+		if strings.Contains(mod.Source, "gke-cluster") || strings.Contains(mod.Source, "pre-existing-gke-cluster") {
```

Comment:
Can we always add kubectl provider to all the blueprints ? What is the downside of this action ?

Reply:
It can lead to an invalid terraform in non-gke case, as kubectl provider depend on gke-cluster / pre-existing-gke-cluster module output, for example.
```
provider "kubectl" {
  apply_retry_count      = 15
  cluster_ca_certificate = module.gke_cluster.cluster_ca_certificate
  host                   = module.gke_cluster.host_endpoint
  load_config_file       = false
  token                  = module.gke_cluster.access_token
}
```

Reply:
Re: downside. Provider plugins are not typically small. They are probably Go which is statically linked. The providers directory for the gke-a3-highgpu.yaml example is 377M in size. I also don't have any idea whether there is a guarantee that they will be available forever so I wouldn't add them unless necessary.

I think Terraform makes you add required_providers explicitly for a reason and we're tripping over it. For now, saying "every GKE cluster triggers the requirement for this provider" seems reasonable. In medium-term, I think the modulereader.ModuleInfo might contain this info.

Reply:
Isn't `gke_cluster` in `module.gke_cluster...` the id that we get from blueprint's block? If yes, I think we cannot assume it's name always will be "gke_cluster". Am I missing something?

Reply:
so we are referring `module.Source` map to module source path and `module.ID` map to user defined unique module id, adding ref to [struct](https://github.com/GoogleCloudPlatform/cluster-toolkit/blob/main/pkg/config/config.go#L230-L240) which hold these value.

Reply:
So the below settings are dependent on user defined module id, right? My point is, we cannot always add the provider because we cannot assume the id for module is always `gke_cluster`.
```
  cluster_ca_certificate = module.gke_cluster.cluster_ca_certificate
  host                   = module.gke_cluster.host_endpoint
  token                  = module.gke_cluster.access_token
```

Reply:
Please let me clarify, had shared that as an example of kubectl provider when `gke-cluster` module id is `gke_cluster`, we have a dynamic setup out here. Please refer this [code block](https://github.com/mohitchaurasia91/cluster-toolkit/blob/fix-gke-module-provider/pkg/config/expand.go#L199-L206).
```
cluster_ca_certificate = module.gke_cluster.cluster_ca_certificate
host                   = module.gke_cluster.host_endpoint
token                  = module.gke_cluster.access_token
```